designs/feature-control.md

```diff
@@ -0,0 +1,85 @@
+
+We will add a feature access control class to the codebase which will allow developers to gate off code paths and allow easily-configurable access to users.
+
+### Success criteria for the addition
+
+* A feature access control system that grants access to unreleased code for development purposes.
+  * Access should not be dependent on a rebuild--this should be user-configurable.
+  * Flags should be undocumented outside the codebase. Even though features will be included in new builds, they will eminently not be ready for general use.
+  * Individual features should be able to be toggled on/off on-the-fly or only on restart ("session-permanent") in a predictable manner.
+  * Tests should be able to inject their own configurations into the feature access control system to ensure correct features are tested.
+
+### Out-of-Scope
+
+* Visual confirmation for a user to discern whether or not a feature is session-permanent when adding a toggle to their `settings.json` file while the extension is running.
+* The ability to pass non-boolean values through feature access control.
+* The ability to permanently override the switch for the feature (as opposed to pruning the toggle and the forking code) when a feature is set to go live.
+
+User Experience Walkthrough
+---------------------------
+
+Developers will be able to gate access to their code with a function. If a corresponding setting is added to the VS Code `settings.json` file, a user will have access to the gated code. This code will then be usable as normal, with no indications that the code is not intended for general use.
+
+Implementation
+--------------
+
+### Design
+
+FeatureAccessControl will be initialized on plugin activation in the `extension.ts:activate()` function. This provides a few things:
+
+* The use of the initialized `DefaultSettingsConfiguration` object, which is also initialized in `activate()`
+* The ability to pass the feature access control object to other objects
+* The ability to lock in session-permanent features as the session is starting.
+
+The FeatureAccessControl object will essentially wrap the SettingsConfiguration object. This wrapping will provide the following:
+
+* A defined list of session-permanent features.
+* Consistent assessment of session-permanent features; initializing the FeatureAccessControl object will lock in the initial values for session-permanent features.
+* On-the-fly assessment of non session-permanent features.
+* An enforced boolean return. Pulling from the VS Code `settings.json` file does not enforce any types.
+* A standard naming convention for feature toggles: searching for feature `featureName` will always correspond to the setting `aws.toggle.featureName`.
+
+Feature access will be granted by passing the initialized FeatureAccessControl object down from the extension activation. Code can then be gated with code similar to the following:
+
+``` typescript
+if (featureController.isFeatureActive('myFeature')) {
```

Comment:
Rather than a string feature name, the `featureController` should expose known methods for each toggle - this will make it much easier to determine where in the codebase a specific toggle is being used (e.g. find references instead of string searching).

Reply:
This is doable. I'll probably add a separate class to handle these toggles; ideally, I don't want users editing the main FeatureController implementation.

If we do decide to have permanent overrides going forward, this is probably the easiest way to implement them.

Reply:
Whether internalizing the string feature name or not, consider using an enum to account for all known feature toggles.